module/zfs/dbuf.c

```diff
@@ -2023,11 +2026,10 @@ dbuf_dirty(dmu_buf_impl_t *db, dmu_tx_t *tx)
 	/*
 	 * If this buffer is already dirty, we're done.
 	 */
-	drp = &db->db_last_dirty;
-	ASSERT(*drp == NULL || (*drp)->dr_txg <= tx->tx_txg ||
+	dr = list_head(&db->db_dirty_records);
+	ASSERT(dr == NULL || dr->dr_txg <= tx->tx_txg ||
 	    db->db.db_object == DMU_META_DNODE_OBJECT);
-	while ((dr = *drp) != NULL && dr->dr_txg > tx->tx_txg)
-		drp = &dr->dr_next;
+	dr_next = dr = dbuf_find_dirty_lte(db, tx->tx_txg);
 	if (dr && dr->dr_txg == tx->tx_txg) {
```

Comment:
I think it would be more clear if we can avoid changing the definition of `dr` (from "the record after the one we want" to "the new record we're inserting").  How about something like:
```
dbuf_dirty_record_t *dr_head = list_head(&db->db_dirty_records);
ASSERT(dr_head...);

dbuf_dirty_record_t *dr_next = dbuf_find_dirty_lte(db, tx->tx_txg);
if (dr_next != NULL && dr_next->dr_txg == tx->tx_txg) {
  // it's not actually the "next" record, it's the one for this txg
   ... return (dr)
}
...
dbuf_dirty_record_t *dr = kmem_zalloc(...);
...
list_insert_before(&db->db_dirty_records, dr_next, dr);
```